### PR TITLE
Add a new diagnostics file class and use it for errors

### PR DESCRIPTION
**Pull request by sauyon**
This was surprisingly easy.

### CODE REVIEW

File: ql/src/semmle/go/Files.qll [hunk 1]
```diff
@@ -266,3 +253,14 @@ class File extends Container, @file, Documentable, ExprParent, GoModExprParent,
 
   override string getAPrimaryQlClass() { result = "File" }
 }
+
+/** A file that has been extracted. */
+class File extends DiagnosticFile {
+  File() {
+    // getAChild is specifically for the Go SAT and so does not apply to non-go files
```
Comment (intrigus-lgtm) [inline ql/src/semmle/go/Files.qll:260]: I assume this is a typo?
If not, what is a "Go SAT"?
```suggestion
    // getAChild is specifically for the Go AST and so does not apply to non-go files
```

File: ql/src/semmle/go/Errors.qll [hunk 2]
```diff
@@ -19,7 +19,7 @@ class Error extends @error {
   int getIndex() { errors(this, _, _, _, _, _, _, _, result) }
 
   /** Gets the file in which this error was reported, if it can be determined. */
-  File getFile() { hasLocationInfo(result.getAbsolutePath(), _, _, _, _) }
+  DiagnosticFile getFile() { hasLocationInfo(result.getAbsolutePath(), _, _, _, _) }
```
Comment (smowton) [inline ql/src/semmle/go/Errors.qll:22]:
```suggestion
  ExtractedOrExternalFile getFile() { hasLocationInfo(result.getAbsolutePath(), _, _, _, _) }
```

File: ql/src/semmle/go/Files.qll [hunk 3]
```diff
@@ -177,25 +177,12 @@ class Folder extends Container, @folder {
...
-      // Exclude `.go` files that have not been extracted. Non-extracted files only exist in the `files`
-      // table if we are reporting compilation errors relating to them in the `diagnostics` table.
-      not path.matches("%.go")
-      or
-      exists(this.getAChild())
-    )
-  }
-
+/** Any file, including files that have not been extracted but must exist as locations for errors. */
+class DiagnosticFile extends Container, @file, Documentable, ExprParent, GoModExprParent,
```
Comment (smowton) [inline ql/src/semmle/go/Files.qll:181]:
```suggestion
/** Any file, including files that have not been extracted but are referred to as locations for errors. */
class ExtractedOrExternalFile extends Container, @file, Documentable, ExprParent, GoModExprParent,
```

File: ql/src/semmle/go/Files.qll [hunk 4]
```diff
@@ -266,3 +253,14 @@ class File extends Container, @file, Documentable, ExprParent, GoModExprParent,
 
   override string getAPrimaryQlClass() { result = "File" }
 }
+
+/** A file that has been extracted. */
+class File extends DiagnosticFile {
```
Comment (smowton) [inline ql/src/semmle/go/Files.qll:258]:
```suggestion
class File extends ExtractedOrExternalFile {
```

**Review by smowton — COMMENTED** [review]
Let's name this for what it is, not its one current usage; otherwise LGTM.

**Review by phetmanyttk — APPROVED** [review]

**Review by smowton — APPROVED** [review]

**Comment by smowton** [comment]
Rebased and merging on green

**Comment by smowton** [comment]
Rebased yet again

**Pull request merged**